lib/propolis/src/attachment.rs

```diff
@@ -0,0 +1,287 @@
+        match &*guard {
+            AttachPoint::Attached(data) => {
+                Some(when_attached(&data.1, &data.0))
+            }
+            AttachPoint::Detached(..) => None,
+        }
+    }
+    pub fn access_fallback<R>(
+        &self,
+        when_attached: impl FnOnce(&T, &AttachState) -> R,
+        fallback: impl FnOnce(&AttachState) -> (),
+    ) -> Option<R> {
+        let guard = self.0.lock().unwrap();
+        match &*guard {
+            AttachPoint::Attached(data) => {
+                Some(when_attached(&data.1, &data.0))
+            }
+            AttachPoint::Detached(state) => {
+                fallback(state);
+                None
+            }
+        }
+    }
+}
+
+enum AttachPoint<T> {
+    Attached(Arc<(AttachState, T)>),
+    Detached(AttachState),
+}
+impl<T> Default for AttachPoint<T> {
+    fn default() -> Self {
+        Self::Detached(AttachState::new())
+    }
+}
+
+/// State of an Attachment ([FrontAttachment] or [BackAttachment])
+#[derive(Clone)]
+pub struct AttachState(Arc<AttachStateInner>);
+impl AttachState {
+    /// Is this Attachment attached to a corresponding peer?
+    ///
+    /// Because the [AttachState] can be cloned to outlive its initial access
+    /// via [FrontAttachment::access] or [BackAttachment::access], the ability
+    /// to check attachedness remains relevant.
+    pub fn is_attached(&self) -> bool {
+        self.0.is_attached.load(Ordering::Acquire)
+    }
+    /// Is the Attachment frontend paused?
+    pub fn is_paused(&self) -> bool {
+        self.0.is_paused.load(Ordering::Acquire)
+    }
+    /// Is the Attachmend backend stopped (from processing additional requests)
+    pub fn is_stopped(&self) -> bool {
+        self.0.is_stopped.load(Ordering::Acquire)
+    }
+    /// Notify waiters of event (either attachment state change, or of
+    /// attachment-specific events, such as new pending requests)
+    pub fn notify(&self) {
+        self.0.notify.notify_waiters()
+    }
+    /// Wait for notification relating to Attachment
+    pub fn notified(&self) -> tokio::sync::futures::Notified<'_> {
+        self.0.notify.notified()
+    }
+    /// Mark Attachment frontend as paused
+    pub fn pause(&self) {
+        if !self.0.is_paused.swap(true, Ordering::Release) {
+            self.0.notify.notify_waiters();
+        }
+    }
+    /// Mark Attachment frontend as resumed (no longer paused)
+    pub fn resume(&self) {
+        if self.0.is_paused.swap(false, Ordering::Release) {
+            self.0.notify.notify_waiters();
+        }
+    }
+    /// Mark Attachment backend as stopped
+    pub fn stop(&self) {
+        if !self.0.is_stopped.swap(true, Ordering::Release) {
+            self.0.notify.notify_waiters();
+        }
+    }
+    /// Mark Attachment backend as started (no longer stopped)
+    pub fn start(&self) {
+        if self.0.is_stopped.swap(false, Ordering::Release) {
+            self.0.notify.notify_waiters();
+        }
+    }
```

Comment:
I would have assumed the `watch` API was more hostile to synchronous callers, but it's mostly reasonable, so that's good to know.

All the same, it'd still require pinging the Condvar for the synchronous waiters, so I think I'm going to leave this as-is for now.  Maybe conditions will dictate otherwise in the future.

Reply:
Synchronous waiters _could_ potentially use a runtime's [`Handle::block_on`](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.block_on) and do `handle.block_on(watch.changed())`...but i think the condvar is probably much more efficient, as the `block_on` would spawn a task. So, I agree, let's carry on with this.